### PR TITLE
adding ability to have dictionary value types

### LINKED ISSUE #3: Hydratable Dictionaries
While working on subscriptions, I found a need to have something like

`
@hy('object') subscriptions: {[key in SubscriptionType]: Subscription};
`

where `Subscriptions` is Hydratable.

I started by just defining a new Hydratable class that looked like:
`
class SubscriptionDict extends Hydratable<{[key in SubscriptionType]: Subscription}> {
  @hy(Subscription) [SubscriptionType.CELLULAR]: Subscription;
  @hy(Subscription) [SubscriptionType.STORAGE]: Subscription;
  @hy(Subscription) [SubscriptionType.LEAK_MONITORING]: Subscription;
  ...
}
`

And that works fine, but it'd be nice if the Hydratable class already handled Dictionary types like this. 
I think the HydratableType would still be 'object' but there'd be a new HydratableOption `dictionaryValueType`.

So it could look like this:
`@hy('object', {dictionaryValueType: Subscriptions}) subscriptions: {[key in SubscriptionType]: Subscription};`

### PR DESCRIPTION
**Pull request by DanielClyde**
Closes #3

### CODE REVIEW

File: src/hydrate.ts [hunk 1]
```diff
@@ -273,6 +274,13 @@ export class Hydratable<T> {
 
   private setObject(key: string, value: any, options: HydrateOptions) {
     if (!options.allowNull && value === null) { return; }
+    const type = options.dictionaryValueType;
+    if (type && type instanceof Function) {
+      Object.keys(value).forEach(key => {
+        value[key] = new type(value[key]);
+      });
+      this.setOnThis(key, this.copyObject(value));
```
Comment (mjclyde) [inline src/hydrate.ts]:
```suggestion
```

Comment (mjclyde) [reply src/hydrate.ts]: Would this change make sense to you? Because this is done down on line 284 anyways.

**Comment by mjclyde** [comment]
Thanks for adding this! I only had one code suggestion. Thank you for adding a test case as well. 👍🏻

File: src/hydrate.ts (hunk 1 above)
Comment (DanielClyde) [reply src/hydrate.ts]: Oh haha good catch! I'll make that change 👍

**Pull request merged**